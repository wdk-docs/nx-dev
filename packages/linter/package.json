{
  "name": "@nrwl/linter",
  "version": "0.0.1",
  "description": "Lint Plugin for Nx",
  "repository": {
    "type": "git",
    "url": "https://github.com/nrwl/nx.git",
    "directory": "packages/linter"
  },
  "keywords": [
    "Monorepo",
    "Web",
    "Lint",
    "TSLint",
    "ESLint",
    "CLI"
  ],
  "main": "./index.js",
  "typings": "./index.d.ts",
  "author": "Victor Savkin",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/nrwl/nx/issues"
  },
  "homepage": "https://nx.dev",
  "ng-update": {
    "requirements": {},
    "migrations": "./migrations.json"
  },
  "builders": "./executors.json",
  "schematics": "./generators.json",
  "peerDependencies": {
    "eslint": "^8.0.0"
  },
  "dependencies": {
    "@phenomnomnominal/tsquery": "4.1.1",
    "@nrwl/devkit": "*",
    "@nrwl/jest": "*",
    "tmp": "~0.2.1",
    "tslib": "^2.3.0"
  },
  "peerDependenciesMeta": {
    "eslint": {
      "optional": true
    }
  }
}
