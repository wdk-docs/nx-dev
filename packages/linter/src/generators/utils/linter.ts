export enum Linter {
  EsLint = 'eslint',
  TsLint = 'tslint',
  None = 'none',
}
