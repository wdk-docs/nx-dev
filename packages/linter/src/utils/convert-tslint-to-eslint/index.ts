export {
  ConvertTSLintToESLintSchema,
  ProjectConverter,
} from './project-converter';

// For testing
export { exampleRootTslintJson } from './example-tslint-configs';
