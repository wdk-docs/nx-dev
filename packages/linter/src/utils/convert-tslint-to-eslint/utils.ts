import { addDependenciesToPackageJson, logger } from '@nrwl/devkit';
import type { Tree, GeneratorCallback } from '@nrwl/devkit';
import type { Linter } from 'eslint';
import type { TSLintRuleOptions } from 'tslint-to-eslint-config';
import { convertTslintNxRuleToEslintNxRule } from './convert-nx-enforce-module-boundaries-rule';
import { convertToESLintConfig } from './convert-to-eslint-config';

export function ensureESLintPluginsAreInstalled(
  host: Tree,
  eslintPluginsToBeInstalled: string[]
): GeneratorCallback {
  if (!eslintPluginsToBeInstalled?.length) {
    return () => undefined;
  }

  const additionalDevDependencies = {};

  for (const pluginName of eslintPluginsToBeInstalled) {
    additionalDevDependencies[pluginName] = 'latest';
  }

  logger.info(
    '\nINFO: To most closely match your tslint.json, we will ensure the `latest` version of the following eslint plugin(s) are installed:'
  );
  logger.info('\n  - ' + eslintPluginsToBeInstalled.join('\n  - '));
  logger.info(
    '\nPlease note, you may later wish to pin these to a specific version number in your package.json, rather than leaving it open to `latest`.\n'
  );

  return addDependenciesToPackageJson(host, {}, additionalDevDependencies);
}

/**
 * We don't want the user to depend on the TSLint fallback plugin, we will instead
 * explicitly inform them of the rules that could not be converted automatically and
 * advise them on what to do next.
 */
function warnInCaseOfUnconvertedRules(
  tslintConfigPath: string,
  unconvertedTSLintRules: TSLintRuleOptions[]
): void {
  const unconvertedTSLintRuleNames = unconvertedTSLintRules
    .filter(
      // Ignore formatting related rules, they are handled by Nx format/prettier
      (unconverted) =>
        !['import-spacing', 'whitespace', 'typedef'].includes(
          unconverted.ruleName
        )
    )
    .map((unconverted) => unconverted.ruleName);

  if (unconvertedTSLintRuleNames.length > 0) {
    logger.warn(
      `\nWARNING: Within "${tslintConfigPath}", the following ${unconvertedTSLintRuleNames.length} rule(s) did not have known converters in https://github.com/typescript-eslint/tslint-to-eslint-config`
    );
    logger.warn('\n  - ' + unconvertedTSLintRuleNames.join('\n  - '));
    logger.warn(
      '\nYou will need to decide on how to handle the above manually, but everything else has been handled for you automatically.\n'
    );
  }
}

export async function convertTSLintConfig(
  rawTSLintJson: any,
  tslintJsonPath: string,
  ignoreExtendsVals: string[]
) {
  const convertedProject = await convertToESLintConfig(
    tslintJsonPath,
    rawTSLintJson,
    ignoreExtendsVals
  );
  convertedProject.convertedESLintConfig.rules ||= {};

  /**
   * Apply the custom converter for the nx-module-boundaries rule if applicable
   */
  const convertedNxRule = convertTslintNxRuleToEslintNxRule(rawTSLintJson);
  if (convertedNxRule) {
    convertedProject.convertedESLintConfig.rules[convertedNxRule.ruleName] =
      convertedNxRule.ruleConfig;
  }

  // Remove the `@typescript-eslint/tslint/config` rule
  if (
    convertedProject.convertedESLintConfig.rules[
      '@typescript-eslint/tslint/config'
    ]
  ) {
    delete convertedProject.convertedESLintConfig.rules[
      '@typescript-eslint/tslint/config'
    ];
  }

  warnInCaseOfUnconvertedRules(
    tslintJsonPath,
    convertedProject.unconvertedTSLintRules
  );

  return convertedProject;
}

export function deduplicateOverrides(
  overrides: Linter.Config['overrides'] = []
) {
  const map = new Map();
  for (const o of overrides) {
    const mapKey: string =
      typeof o.files === 'string' ? o.files : o.files.join(',');
    const existing: Set<Linter.ConfigOverride> = map.get(mapKey);
    if (existing) {
      existing.add(o);
      map.set(mapKey, existing);
      continue;
    }
    const set = new Set();
    set.add(o);
    map.set(mapKey, set);
  }

  let dedupedOverrides = [];

  for (const [, overrides] of map.entries()) {
    const overridesArr = Array.from(overrides);
    if (overridesArr.length === 1) {
      dedupedOverrides = [...dedupedOverrides, ...overridesArr];
      continue;
    }
    let mergedOverride = {};
    for (const o of overridesArr) {
      mergedOverride = {
        ...mergedOverride,
        ...(o as any),
      };
    }
    dedupedOverrides.push(mergedOverride);
  }

  return dedupedOverrides;
}
