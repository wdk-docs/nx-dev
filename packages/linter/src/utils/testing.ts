import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { join } from 'path';
import { Tree } from '@angular-devkit/schematics';

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/linter-migrations',
  join(__dirname, '../../migrations.json')
);

export function runMigration<T = any>(
  migrationName: string,
  options: T,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync<T>(migrationName, options, tree)
    .toPromise();
}
