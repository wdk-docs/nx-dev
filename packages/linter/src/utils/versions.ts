export const nxVersion = '*';

export const tslintVersion = '~6.1.0';
export const tslintToEslintConfigVersion = '^2.4.0';
export const buildAngularVersion = '~13.1.0';

export const typescriptESLintVersion = '~5.3.0';
export const eslintVersion = '8.2.0';
export const eslintConfigPrettierVersion = '8.1.0';
