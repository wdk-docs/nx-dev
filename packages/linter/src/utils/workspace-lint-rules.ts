import { joinPathFragments } from '@nrwl/devkit';

export const workspaceLintPluginDir = joinPathFragments('tools/eslint-rules');
