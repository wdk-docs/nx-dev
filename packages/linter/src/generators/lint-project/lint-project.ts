import {
  writeJson,
  updateProjectConfiguration,
  offsetFromRoot,
  readProjectConfiguration,
  formatFiles,
} from '@nrwl/devkit';
import type { ProjectConfiguration, Tree } from '@nrwl/devkit';
import { join } from 'path';
import { Linter } from '../utils/linter';
import { lintInitGenerator } from '../init/init';

interface LintProjectOptions {
  project: string;
  linter?: Linter;
  eslintFilePatterns?: string[];
  tsConfigPaths?: string[];
  skipFormat: boolean;
  setParserOptionsProject?: boolean;
}

function createTsLintConfiguration(
  tree: Tree,
  projectConfig: ProjectConfiguration
) {
  writeJson(tree, join(projectConfig.root, `tslint.json`), {
    extends: `${offsetFromRoot(projectConfig.root)}tslint.json`,
    // Include project files to be linted since the global one excludes all files.
    linterOptions: {
      exclude: ['!**/*'],
    },
    rules: {},
  });
}

function createEsLintConfiguration(
  tree: Tree,
  projectConfig: ProjectConfiguration,
  setParserOptionsProject: boolean
) {
  writeJson(tree, join(projectConfig.root, `.eslintrc.json`), {
    extends: [`${offsetFromRoot(projectConfig.root)}.eslintrc.json`],
    // Include project files to be linted since the global one excludes all files.
    ignorePatterns: ['!**/*'],
    overrides: [
      {
        files: ['*.ts', '*.tsx', '*.js', '*.jsx'],
        /**
         * NOTE: We no longer set parserOptions.project by default when creating new projects.
         *
         * We have observed that users rarely add rules requiring type-checking to their Nx workspaces, and therefore
         * do not actually need the capabilites which parserOptions.project provides. When specifying parserOptions.project,
         * typescript-eslint needs to create full TypeScript Programs for you. When omitting it, it can perform a simple
         * parse (and AST tranformation) of the source files it encounters during a lint run, which is much faster and much
         * less memory intensive.
         *
         * In the rare case that users attempt to add rules requiring type-checking to their setup later on (and haven't set
         * parserOptions.project), the executor will attempt to look for the particular error typescript-eslint gives you
         * and provide feedback to the user.
         */
        parserOptions: !setParserOptionsProject
          ? undefined
          : {
              project: [`${projectConfig.root}/tsconfig.*?.json`],
            },
        /**
         * Having an empty rules object present makes it more obvious to the user where they would
         * extend things from if they needed to
         */
        rules: {},
      },
      {
        files: ['*.ts', '*.tsx'],
        rules: {},
      },
      {
        files: ['*.js', '*.jsx'],
        rules: {},
      },
    ],
  });
}

export async function lintProjectGenerator(
  tree: Tree,
  options: LintProjectOptions
) {
  const installTask = lintInitGenerator(tree, {
    linter: options.linter,
  });
  const projectConfig = readProjectConfiguration(tree, options.project);

  if (options.linter === Linter.EsLint) {
    projectConfig.targets['lint'] = {
      executor: '@nrwl/linter:eslint',
      outputs: ['{options.outputFile}'],
      options: {
        lintFilePatterns: options.eslintFilePatterns,
      },
    };
    createEsLintConfiguration(
      tree,
      projectConfig,
      options.setParserOptionsProject
    );
  } else {
    projectConfig.targets['lint'] = {
      executor: '@angular-devkit/build-angular:tslint',
      options: {
        tsConfig: options.tsConfigPaths,
        exclude: ['**/node_modules/**', `!${projectConfig.root}/**/*`],
      },
    };
    createTsLintConfiguration(tree, projectConfig);
  }

  updateProjectConfiguration(tree, options.project, projectConfig);

  if (!options.skipFormat) {
    await formatFiles(tree);
  }

  return installTask;
}
