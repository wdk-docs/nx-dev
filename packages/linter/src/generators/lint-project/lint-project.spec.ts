import {
  addProjectConfiguration,
  readProjectConfiguration,
  Tree,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';

import { Linter } from '../utils/linter';
import { lintProjectGenerator } from './lint-project';

describe('@nrwl/linter:lint-project', () => {
  let tree: Tree;

  const defaultOptions = {
    skipFormat: false,
  };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    addProjectConfiguration(tree, 'test-lib', {
      root: 'libs/test-lib',
      targets: {},
    });
  });

  describe('--linter', () => {
    describe('eslint', () => {
      it('should generate a eslint config', async () => {
        await lintProjectGenerator(tree, {
          ...defaultOptions,
          linter: Linter.EsLint,
          eslintFilePatterns: ['**/*.ts'],
          project: 'test-lib',
          setParserOptionsProject: false,
        });

        expect(
          tree.read('libs/test-lib/.eslintrc.json', 'utf-8')
        ).toMatchSnapshot();
      });

      it('should configure the target in workspace.json', async () => {
        await lintProjectGenerator(tree, {
          ...defaultOptions,
          linter: Linter.EsLint,
          eslintFilePatterns: ['**/*.ts'],
          project: 'test-lib',
          setParserOptionsProject: false,
        });

        const projectConfig = readProjectConfiguration(tree, 'test-lib');
        expect(projectConfig.targets.lint).toMatchInlineSnapshot(`
          Object {
            "executor": "@nrwl/linter:eslint",
            "options": Object {
              "lintFilePatterns": Array [
                "**/*.ts",
              ],
            },
            "outputs": Array [
              "{options.outputFile}",
            ],
          }
        `);
      });
    });

    describe('tslint', () => {
      it('should generate a tslint config', async () => {
        await lintProjectGenerator(tree, {
          ...defaultOptions,
          linter: Linter.TsLint,
          tsConfigPaths: ['tsconfig.json'],
          project: 'test-lib',
          setParserOptionsProject: false,
        });

        expect(
          tree.read('libs/test-lib/tslint.json', 'utf-8')
        ).toMatchSnapshot();
      });

      it('should configure the target in workspace.json', async () => {
        await lintProjectGenerator(tree, {
          ...defaultOptions,
          linter: Linter.TsLint,
          tsConfigPaths: ['tsconfig.json'],
          project: 'test-lib',
          setParserOptionsProject: false,
        });

        const projectConfig = readProjectConfiguration(tree, 'test-lib');
        expect(projectConfig.targets.lint).toMatchInlineSnapshot(`
          Object {
            "executor": "@angular-devkit/build-angular:tslint",
            "options": Object {
              "exclude": Array [
                "**/node_modules/**",
                "!libs/test-lib/**/*",
              ],
              "tsConfig": Array [
                "tsconfig.json",
              ],
            },
          }
        `);
      });
    });
  });
});
