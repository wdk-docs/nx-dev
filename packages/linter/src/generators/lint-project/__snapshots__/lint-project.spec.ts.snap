// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nrwl/linter:lint-project --linter eslint should generate a eslint config 1`] = `
"{
  \\"extends\\": [
    \\"../../.eslintrc.json\\"
  ],
  \\"ignorePatterns\\": [
    \\"!**/*\\"
  ],
  \\"overrides\\": [
    {
      \\"files\\": [
        \\"*.ts\\",
        \\"*.tsx\\",
        \\"*.js\\",
        \\"*.jsx\\"
      ],
      \\"rules\\": {}
    },
    {
      \\"files\\": [
        \\"*.ts\\",
        \\"*.tsx\\"
      ],
      \\"rules\\": {}
    },
    {
      \\"files\\": [
        \\"*.js\\",
        \\"*.jsx\\"
      ],
      \\"rules\\": {}
    }
  ]
}
"
`;

exports[`@nrwl/linter:lint-project --linter tslint should generate a tslint config 1`] = `
"{
  \\"extends\\": \\"../../tslint.json\\",
  \\"linterOptions\\": {
    \\"exclude\\": [
      \\"!**/*\\"
    ]
  },
  \\"rules\\": {}
}
"
`;
